com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/CCWindow/CCWindow_cpp.cgt

```diff
@@ -0,0 +1,253 @@
+
+@include "../pyspltuple.cgt"
+<%
+ # Select the Python wrapper function
+ my $pyoutstyle = splpy_tuplestyle($model->getOutputPortAt(0));
+ 
+ # Temporary, until more input types are supported. At the moment, only pass by reference types 
+ # are supported.
+ my $in_pywrapfunc =  $pystyle . '_in__object_out';
+ my $out_pywrapfunc=  'pickle_in__' . $pyoutstyle . '_out';
+%>
+
+#define SPLPY_TUPLE_MAP(f, v, r, occ) \
+    streamsx::topology::Splpy::pyTupleMap(f, v, r)
+
+// Constructor
+MY_OPERATOR::MY_OPERATOR() :
+   funcop_(NULL),
+   pyInNames_(NULL),
+   occ_(-1),
+   window_(*this, 0,
+   CountWindowPolicy(<%=$model->getParameterByName("evictConfig")->getValueAt(0)->getSPLExpression()%>),
+   CountWindowPolicy(<%=$model->getParameterByName("triggerConfig")->getValueAt(0)->getSPLExpression()%>))
+{
+    window_.registerOnWindowTriggerHandler(this);
+    window_.registerAfterTupleEvictionHandler(this);
+
+    const char * in_wrapfn = "<%=$in_pywrapfunc%>";
+    const char * out_wrapfn = "<%=$out_pywrapfunc%>";
+    const char * functions_module = "streamsx.topology.functions";
+    const char * identity_function = "identity";
+
+<%
+# If occ parameter is positive then pass-by-ref is possible
+# Generate code to allow pass by ref but only use when
+# not connected to a PE output port.
+
+ my $oc = $model->getParameterByName("outputConnections");
+
+ if ($oc) {
+    my $occ = $oc->getValueAt(0)->getSPLExpression();
+    if ($occ > 0) {
+        my $pybyrefwrapfunc = 'pickle_in__object_out';
+%>
+
+#undef SPLPY_TUPLE_MAP
+#define SPLPY_TUPLE_MAP(f, v, r, occ) \
+    streamsx::topology::Splpy::pyTupleMapByRef(f, v, r, occ)
+
+    if (!this->getOutputPortAt(0).isConnectedToAPEOutputPort()) {
+       // pass by reference
+       out_wrapfn = "<%=$pybyrefwrapfunc%>";
+       occ_ = <%=$occ%>;
+    }
+<%
+    } 
+ }
+%>
+
+    funcop_ = new SplpyFuncOp(this, out_wrapfn);
+    spl_in_object_out = new SplpyFuncOp(this, in_wrapfn, functions_module, identity_function);
+
+<% if ($pystyle eq 'dict') { %>
+     SplpyGIL lock;
+     pyInNames_ = streamsx::topology::Splpy::pyAttributeNames(
+               getInputPortAt(0));
+<% } %>
+}
+
+// Destructor
+MY_OPERATOR::~MY_OPERATOR() 
+{
+<% if ($pystyle eq 'dict') { %>
+    if (pyInNames_) {
+      SplpyGIL lock;
+      Py_DECREF(pyInNames_);
+    }
+<% } %>
+
+  delete funcop_;
+  delete spl_in_object_out;
+}
+
+// Notify pending shutdown
+void MY_OPERATOR::prepareToShutdown() 
+{
+    funcop_->prepareToShutdown();
+}
+
+// Tuple processing for non-mutating ports
+void MY_OPERATOR::process(Tuple const & tuple, uint32_t port)
+{
+  IPort0Type const &ip = static_cast<IPort0Type const &>(tuple);
+
+<%
+print splpy_inputtuple2value($pystyle);
+%>
+
+<%if ($pystyle eq 'dict') {%>
+@include "../pyspltuple2dict.cgt"
+<%}%>
+
+  PyObject *python_value;
+
+// If the input style is pickle,
+{
+  // None of the streamsx::topology methods in this scope grab the lock
+  // so we need to do it here.
+  SplpyGIL lock; 

```

Comment:
If the input is already a Python object are we grabbing the GIL for no reason?

Reply:
Fixed. The GIL should now only be grabbed when necessary.